Vim performance scope 2

kbrock opened the pull request:
This extracts `VmPerformanceDaily.find_entries` into a helper method that is also using a scope. This is for rbac scopes

/cc @matthewd @Fryguy 
/cc @gtanzillo FYI - thanks for help with other sections

kbrock commented:
unWIPping - fixed rbac warnings, added a number of helpers to reduce the "if it is daily" logic. There is still more to do there

matthewd commented on app/models/metric/helper.rb:
```diff
@@ -64,6 +72,19 @@ def self.next_daily_timestamp(ts, tz = nil)
+  # @return Range<DateTime,DateTime>
+  def self.timestamp_from_hash(range)
+    return range unless range.kind_of?(Hash)
+    range[:days] ||= 20
```
This should probably just use a local: there's no reason for us to be mutating the hash we were passed. Even if we're planning to kill off this hash (are we?), in the meantime it'll be slightly easier to reason about without hidden side-effects.

chessbyte commented on app/models/metric/helper.rb:
```diff
@@ -1,6 +1,14 @@
+    interval_name == "realtime" ? [Metric, :metrics] : [rollup_class, :metric_rollups]
+  end
+
+  def self.find_for_interval_name(interval_name, rollup_class = nil, time_profile = nil, tz = nil, timestamp = nil)
```
would the optional parms be clearer as a hash?

kbrock commented:
@chessbyte I updated `find_for_interval_name` - there was only 1 optional parameter.

But your primary question is still present.]
Should those go in as a hash or as multiple methods.

The goal is to mimic [`find_all_by_interval_and_time_range`](http://github.com/ManageIq/manageiq/blob/master/app/models/metric_rollup.rb#L4) as closely as possible

chessbyte commented:
@gtanzillo @blomquisg @Fryguy please review

matthewd commented on app/services/container_dashboard_service.rb:
```diff
@@ -229,11 +229,8 @@ def daily_network_metrics
-                                          .where('timestamp > ?', 30.days.ago.utc).order("timestamp")
-    end
-    @daily_metrics
+    @daily_metrics ||= Metric::Helper.find_for_interval_name('daily', nil, nil, @controller.current_user.get_timezone)
```
Using kwargs for all but the first parameter would avoid this `nil, nil` -- I think that's what @chessbyte was suggesting.

kbrock replied:
Every call but 2 use all of the args when calling `find_for_interval_name`. One skips `timestamp`, as we want to pass a non-range. This is the other instance, that skips these 2 args in the middle.

Using kwargs seems like over kill. On the other hand, 4 args seems a bit much as well. So if it increases readability, then sounds good to me.

@chessbyte were you thinking kwargs?

chessbyte replied:
@kbrock correct.  I just did not know the technical term `kwargs` :-)
@matthewd thanks for reading my mind.

kbrock commented on app/models/miq_report/generator/trend.rb:
```diff
@@ -24,33 +24,24 @@ def build_results_for_report_trend(options)
+      recs = Metric::Helper.find_for_interval_name("daily", trend_klass, time_profile, tz)
+                           .where(where_clause).where(:timestamp => timestamp).includes(includes)
     else
       start_time = Time.now.utc - db_options[:start_offset].seconds
```
@gtanzillo I noticed that we use different `:start_offset` logic here from `Metric::Helper.timestamp_from_offset`. Do you have an idea if this was intentional? This one does `Time.now.utc - db_options[:start_offset].seconds` where the others seem to subtract from 11pm last night.

gtanzillo replied:
@kbrock I believe that the difference in logic has to do with whether the report interval is daily vs. hourly. For daily reports we want to have the start_time be at the beginning of the day in the specified time zone so the calculation starts at the end of the previous day (so that the first hour of the day would be the first selected). For an hourly report, the day does not matter so it will just subtract the offset from the current time UTC.

kbrock replied:
@gtanzillo thanks. seems 90%(?) of the calls use the previous day logic. Would it be ok to push the "hourly" stuff towards that? Or do you think that would possibly not convey the user's intent/expectation?

gtanzillo replied:
@kbrock I think (but not really sure) that it may be important for hourly trend reports to have the offset start from the current hour and not the beginning of the day. How about moving the hourly logic to `get_time_range_from_offset` with an option arg for `interval` with a default of `:daily`?

kbrock replied:
agreed - I'll leave as is

kbrock commented on app/models/metric/helper.rb:
```diff
@@ -3,6 +3,15 @@ def self.class_and_association_for_interval_name(interval_name)
     interval_name == "realtime" ? [Metric, :metrics] : [MetricRollup, :metric_rollups]
   end
 
+  def self.find_for_interval_name(interval_name, rollup_class, time_profile, tz)
```
would prefer not to make this kwargs

How about:

```ruby
 def self.find_for_interval_name(interval_name, time_profile_or_tz, rollup_class = nil)
 end
```

chessbyte commented on app/models/metric/helper.rb:
```diff
@@ -154,16 +176,16 @@ def self.get_time_zone(options)
+  def self.time_range_from_offset(start_offset, end_offset, tz)
     # Get yesterday at 23:00 in specified TZ. Then convert that to UTC. Then subtract offsets
-    tz = Metric::Helper.get_time_zone(options)
+    tz ||= Metric::Helper.get_time_zone(nil)
```
why pass nil to `get_time_zone`?  why not simply `tz ||= Metric::Helper.get_time_zone`?

kbrock replied:
good one

chessbyte commented on app/models/miq_report/generator.rb:
```diff
@@ -233,10 +233,10 @@ def _generate_table(options = {})
       end
 
-      start_time, end_time = Metric::Helper.get_time_range_from_offset(db_options[:start_offset], db_options[:end_offset], :tz => tz)
+      timestamp = Metric::Helper.time_range_from_offset(db_options[:start_offset], db_options[:end_offset], tz)
```
suggest renaming local variable `timestamp` to `time_range` for clarity.

chessbyte commented on app/models/miq_report/generator/trend.rb:
```diff
@@ -28,10 +28,10 @@ def build_results_for_report_trend(options)
     if db_options[:interval] == "daily"
       includes = include.blank? ? [] : include.keys
-      start_time, end_time = Metric::Helper.get_time_range_from_offset(db_options[:start_offset], db_options[:end_offset], :tz => tz)
+      timestamp = Metric::Helper.time_range_from_offset(db_options[:start_offset], db_options[:end_offset], tz)
```
suggest renaming local variable `timestamp` to `time_range` for clarity.

kbrock commented:
ok.

1. changed `timestamp` to `time_range
2. reworked `time_profile` / `tz` so we didn't have those funky nil parameters.

Let me know what you think
